examples/WebRadioChunkedDemo/WebRadioChunkedDemo.ino

```diff
@@ -0,0 +1,225 @@
+/***
+ * Removes the chunk control data from the helper buffer.
+ *
+ * Only the following chunk control bytes are removed:
+ * \r\n<byte><byte>\r\n
+ * \r\n<byte><byte><byte>\r\n
+ * \r\n<byte><byte><byte><byte>\r\n
+ *
+ */
+void remove_chunk_control_data_from_helper_buffer() {
+  if (helperBuffer[0] != '\r') {
+    // die fast
+    return;
+  }
+  if (helperBuffer[1] != '\n') {
+    // die fast
+    return;
+  }
+
+  if (helperBuffer[4] == '\r' && helperBuffer[5] == '\n') {
+    // 6 bytes length chunk control section discovered
+    // \r\n<byte><byte>\r\n
+    helperBufferCount = 2;
+    Serial.println("Removed control data: 6 bytes");
+
+    return;
+  }
+
+  if (helperBuffer[5] == '\r' && helperBuffer[6] == '\n') {
+    // 7 bytes length chunk control section discovered
+    // \r\n<byte><byte><byte>\r\n
+    helperBufferCount = 1;
+    Serial.println("Removed control data: 7 bytes");
+
+    return;
+  }
+
+  if (helperBuffer[6] == '\r' && helperBuffer[7] == '\n') {
+    // 8 bytes length chunk control section discovered
+    // \r\n<byte><byte><byte><byte>\r\n
+    helperBufferCount = 0;
+    Serial.println("Removed control data: 8 bytes");
+
+    return;
+  }
+}
+
+/***
+ * Puts a byte to the input of helper buffer and returns a byte from the output of helper buffer.
+ * In the meantime it tries to remove the chunk control data (if any available) from the buffer.
+ *
+ * @param incoming byte of the audio stream
+ * @return outgoing byte of the audio stream (if available) or -1 when no bytes available (this happens
+ * when the buffer is being populated with the data or after removal of chunk control data)
+ */
+int16_t put_through_helper_buffer(uint8_t newValue) {
+  //
+  int16_t result = -1;
+
+  if (helperBufferCount == 8) {
+    result = helperBuffer[0];
+  }
+
+  helperBuffer[0] = helperBuffer[1];
+  helperBuffer[1] = helperBuffer[2];
+  helperBuffer[2] = helperBuffer[3];
+  helperBuffer[3] = helperBuffer[4];
+  helperBuffer[4] = helperBuffer[5];
+  helperBuffer[5] = helperBuffer[6];
+  helperBuffer[6] = helperBuffer[7];
+  helperBuffer[7] = newValue;
+
+  helperBufferCount++;
+  if (helperBufferCount > 8) {
+    helperBufferCount = 8;
+  }
+
+  remove_chunk_control_data_from_helper_buffer();
+
+  return result;
+}
+
+/***
+ * Removes the chunk control data from the input audio stream. Data are written back to the input
+ * buffer and the number of bytes available (after processing) is returned.
+ *
+ * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding
+ * @see https://en.wikipedia.org/wiki/Chunked_transfer_encoding
+ *
+ * @param data a pointer to the input buffer of audio stream
+ * @param length a number of input bytes to be processed
+ * @return the number of available bytes in the input buffer after removing the chunk control data
+ */
+uint8_t remove_chunk_control_data(uint8_t *data, size_t length) {
+
+  uint8_t writeindex = 0;
+  uint8_t index = 0;
+  for (index = 0; index < lengt; index++) {

```

Comment:
@philippedc pointed out that there is a bug: `length` instead of `lengt`.
He also said that: "by the way, if the code gives sound, the quality is awful with much more glitches than ever".
I wonder whether it's due to reducing the buffer implementation or something was changed in the code omitting glitches?

Reply:
Yes, my bad. I have fixed the typo in the ```lengt``` method signature only. Sorry.

I have no idea why it doesn;t work for @philippedc. For me it works perfect, no more glitches. I will check that again.

Reply:
That is very strange...
I have : Arduino IDE last release 1.8.12, core ESP8266 2.4.2 and 2.6.3 (my web radio doesn't compile with 2.6.3 due to the web server library conflict) 
IwIP variant : v2 higher bandwich
CPU : 80MHz or 160MHz


Here are 2 stations that use to glitch a lot. In fact they send many chunk information....
Can you test them with your code ?
23.82.11.89:30228/stream
23.82.11.88:5128/stream

Reply:
@philppedc,

>Can you test them with your code ?
23.82.11.89:30228/stream
23.82.11.88:5128/stream

Those two URLs work nice with my code. There is no glitches at all, the sound plays nice. I have used the sam settings as you: 
````
core ESP8266 2.4.2
IwIP variant : v2 higher bandwidth
CPU : 80MHz
````

Does your ESP8266 also hosts a webserver inside? Maybe it has also some impact, so the sound has glitches and is brassy, because the processor also need to do the things related to the webserver?
My test code is simple; I do not have any webserver, so it only reads the stream, remove chunks data information and sends the bytes to VS1053.

Reply:
> Does your ESP8266 also hosts a webserver inside?

I've tested without and with. No special issue with.
I'm wondering if it is not a matter of cutting length: when my code cuts 5 bytes yours cuts 7 - it is what I notice on the console.

Reply:
You are right @philippedc, my code is not able to cut 5 bytes because it was not designed to cut such a short data blocks - this is the issue here and that explains why you still have glitches. Probably you have different wifi conditions and server decides to send you a shorter chunks.
I will fix the code an let you know to test again.

Reply:
@philippedc, I have commited additional fix. Please retest.

Reply:
@wmarkow sorry I do not notice any difference. Also it still removes 7 bytes when it should be 5.
I explain what happens during listening:
1- new station choosen: it is playing
2- Immediate console message : Removed control data: 8 bytes
3- 2 seconds later (arround) : a glitch
4- 2 seconds later (arround) : console message : Removed control data: 8 bytes
5- 2 seconds later (arround) : a glitch
6- 2 seconds later (arround) : console message : Removed control data: 8 bytes
7- 2 seconds later (arround) : a glitch
8- then it accelerates with at least 3 glitches and message console per second
the message console says: Removed control data: 7 bytes
9- after a dozen of seconds the webradio aborts
Please note I have the same sequence with my code: the marker remove is slow at the beginning, and after few seconds it is about 3 per seconds. 
The difference is that I cut 6 bytes when you cut 8, and 5 when it is 7 for you.

Reply:
Ok, I see. Now it is time when I use your code in my tests. Could you post it here? Or is it the one posted [here](https://github.com/baldram/ESP_VS1053_Library/issues/47#issuecomment-601866643)?

Reply:
@wmarkow, yes it is this one. Here again:

````
void loop() {

    ArduinoOTA.handle();        // allow update on air
    server.handleClient();      // call to the server
    static byte tempo = 0;
    tempo++;
 
// if signal is lost
  if( !client.connected() && tempo%10 == 0 ) {
    WiFi.status();            // verify wifi
    Serial.println("Tentative de reconnexion...");
    if( client.connect(urlStation, portStation.toInt()) ) {
      Serial.println("Tentative d'accès...");
      client.print(
			String("GET ") + path + " HTTP/1.1\r\n" + "Host: " + host + "\r\n"
					+ "Icy-MetaData: 0" + "\r\n" + "Connection: close\r\n\r\n");
    }
  }  // end of !client.connected() test
  else {
    
// if station is connected :)
    byte i=0;
    byte noise=0;

    if( client.available() && !pause ) {
      while( i < bytesread ) {
        mp3buff[i] = client.read();
        //delayMicroseconds(10); 
      
        if( i>noise && mp3buff[i] == '\n' && mp3buff[i-1] == '\r' ) {
          if( noise == 0 ) noise =i;
          else if( (i-noise)<10 ) {
            Serial.print(i); Serial.print(":"); Serial.println(i-noise);  
            i=i-(i-noise);
          }
          i--;
        }
        else i++;
      }    // end of while
    
      player.playChunk(mp3buff, bytesread); 
      //yield();
      z=0;
    }      // end of client.available
    yield();

  }  // end of else
}    // end of loop
````

Reply:
@philippedc, I have tested your code; it works on my test environment; not all glitches are removed but it is better than nothing.
You also write:
>The difference is that I cut 6 bytes when you cut 8, and 5 when it is 7 for you.

I'm not quite sure how you find out that your code cuts 6 bytes or 5 bytes. I have the feeling that you always cut all **\r\n** sequences, which is always 2 bytes length. Please correct me.


I have retested also my **Web radio chunked demo** with **Arduino IDE 1.8.12**. I tought that maybe there is a difference when compiling in Arduino IDE or in Sloeber/Eclipse (theoretically there may be some compilation flags differences). With Arduino IDE it also works perfect for me; no glitches at all.
I'm running out of ideas why it doesn't work for you :(

Reply:
@philippedc, [here](https://github.com/baldram/ESP_VS1053_Library/issues/47#issuecomment-601432249) you posted some results from your algorithm. The print outs are in format:
**i:i-noise**
so for example the output **31:4**
gives
i = 31
i-noise = 4, so noise = i - 4 = 31 - 4 = 27
This indicates that you have removed the following sequence
````
i: 26 27 28 29 30 31
   \r \n 0x 0x \r \n
````
so you have removed 6 bytes. My algorithm should also remove the same 6 bytes in this case.

Reply:
Hi @wmarkow I've tested again your code, but I definitely have the sound trunk, like the chunk cut is not of the right size. Moreover the code aborts after a while.... I've tried to follow your algorithm, it is not easy....

However I'm still investigating and I have notice something strange. According to https://en.wikipedia.org/wiki/Chunked_transfer_encoding the chunk is like the format:
````
4\r\n
Wiki\r\n
5\r\n
pedia\r\n
E\r\n
 in\r\n
\r\n
chunks.\r\n
0\r\n
\r\n
```
for a chunk size of 4, 5, 14 and 0. And a \r\n is always followed by another one after few bytes.

So I have changed a little bit my code for the following:
````
    
// if station is connected :)
    int i=0;
    int noise=0;

    if( client.available() && !pause ) {
      while( i < bytesread ) {
        mp3buff[i] = client.read();
        
        if( i>noise && mp3buff[i] == '\n' && mp3buff[i-1] == '\r' ) {
          i--;
          Serial.println(i);
          
          if( noise == 0 ) noise =i;
          else if( (i-noise)<10 ) {
            Serial.print(noise); Serial.print(":"); Serial.println(i);  
            i = i-(i-noise);
            noise = 0;
          }
        }
        else i++;
      }    // end of while
    
      player.playChunk(mp3buff, bytesread); 
    }      // end of client.available
    yield();

  }  // end of else
}    // end of loop
````
just to have a look after all \r\n sequence. What I have noticed is that every glitch is due to a  \r \n  sequence not followed by another  \r \n  after few bytes! It is an alone \r\n .... I've tried to delete these alone  \r \n with the  i-- but I still listen the glitch. So there must be something else to suppress. But what ?
Please see the console display:

![image](https://user-images.githubusercontent.com/50800330/77806960-8ce57d00-7086-11ea-9028-7a1878759810.png)

and note the "36" alone.

I am surprised that you do not hear these alone \r\n glitches with your code, do you confirm ?

Reply:
@ohillipedc, indeed - I think that my code removes all of the glitches related to **\r\n**. My code removes all chunk control data that are with the format as below:
````
\r\n<byte>\r\n
\r\n<byte><byte>\r\n
\r\n<byte><byte><byte>\r\n
\r\n<byte><byte><byte><byte>\r\n
````

It removes minimal 5 bytes and maximal 8 bytes. From your Wikipedia example, the **\r\n
5\r\n** sequence will be removed. I just have noticed that the Wikipedia example show also that **\r\n** are sent as a chunkd data, not as a chunk control block. Here:
````
pedia\r\n
E\r\n
 in\r\n
\r\n
chunks.\r\n
````
My algorithm will fail here because it will remove the following parts of the stream:
````
\r\n
E\r\n
````
So in this case some glitches are theoretically possible but from the other side what is the probability that a sequence of **\r\n** will be really a part of audio stream?


>but I still listen the glitch. So there must be something else to suppress. But what ?

Maybe it is because you analyze the stream in the 32 bytes chunks. When the leading **\r\n** is somewhere at the end of your 32 bytes chunk and the ending **\r\n** is somewhere at the beginning of the next 32 bytes chunk, then your algorithm may not cut it correctly. 
My solution has an 8 bytes length middle buffer which is shifted constantly and the algorithm looks for those chunk contorl patterns also constantly.

Reply:
> Maybe it is because you analyze the stream in the 32 bytes chunks. When the leading **\r\n** is somewhere at the end of your 32 bytes chunk and the ending **\r\n** is somewhere at the beginning of the next 32 bytes chunk, then your algorithm may not cut it correctly.
> My solution has an 8 bytes length middle buffer which is shifted constantly and the algorithm looks for those chunk contorl patterns also constantly, so I do not analyse the stream in chunks; the stream is analysed continuously.

I see, in fact if screen capture has been done with a buffer of 32 bytes, it is working identically with a buffer of 2048 bytes. 
May be the glitch is due by a chunk at the high end of the buffer. But that does not answer why I find an alone \r\n in the middle of the stream. I do not have many glitches, average of one every 5 to 10 seconds so it is easy to verify the console.
However you're right with the idea of a shift buffer, I will work on it.

Reply:
OK, so the chunk information is under the format 'number_of_bytes'\r\n. I understand that the preliminary \r\n is optional, and not count in the chunk, right ?
So I did the following:
`

    int i=0;
    int bytesread = BUFFERSIZE;    // = 4096
    static int counter = 0;
    int chunk;

    if( client.available()) {
      while( i < bytesread ) {
        mp3buff[i] = client.read();

        if( i>1 && mp3buff[i] == '\n' && mp3buff[i-1] == '\r' ) {
          chunk = mp3buff[i-2];
            
          Serial.print(i); Serial.print(":"); Serial.print(chunk); Serial.print(":"); Serial.println(counter);
          counter = 0;
        }
        i++;
        counter++;
      }    // end of while
    
      player.playChunk(mp3buff, bytesread); 
    }      // end of client.available
    yield();

  }  // end of else
}    // end of loop
`
which means that each time "I" see a \r\n sequence I read the past byte: if i=\n and i-1=\r so i-2=chunk_size.
Here is the result:
![image](https://user-images.githubusercontent.com/50800330/77966274-f95cb800-72e2-11ea-8db7-82bcf5912122.png)
1st column => i
2nd column => chunk_size
3rd column => value of the byte counter 
I well can see the counter indicates 5 every 2 lines: it is the count of the 5 bytes with the sequence \r\nX\r\n, but I do not understand the correlation between this X value that should be the chunk size and the result of the counter. WHat is wrong ?

Reply:
.... or does that mean when a receive a packet of 1346 bytes, only the first 48 are useful for the streaming ???

Reply:
I've find the way to solve all my problems. see https://github.com/baldram/ESP_VS1053_Library/issues/52
Many thanks for your help !!!!!

Reply:
>OK, so the chunk information is under the format 'number_of_bytes'\r\n. I understand that the preliminary \r\n is optional, and not count in the chunk, right ?

Preliminary \r\n is not optional, I think. The whole format of the stream is a s below:
```
metadata_about_the_stream\r\n
ZZZZ\r\n
chunk\r\n
ZZZZ\r\n
chunk\r\n
ZZZ\r\n
chunk\r\n
ZZ\r\n
chunk\r\n
Z\r\n
...
````
where ZZZZ or ZZZ or ZZ  or Z is the length of the next chunk data. But attention, it looks like this length is encoded in ASCII HEX (human readable), so if ZZZZ is **a765** then the legth of chunks is **0xa765** or if Z is **5** then the chunk length is 0x5 which is exactly five bytes. If you decode a chunk length byte as **56 DEC** then according to the ASCII table it is a digit **8**, so the chunk length is **8** bytes (not 56 bytes). 

If you take a look again at the format above, you will notice a pattern in the stream. All what needs to be cut out is **\r\nZZZZ\r\n** or **\r\nZZZ\r\n** or **\r\nZZ\r\n** or **\r\nZ\r\n**, because these are the chunk control data; everything else what is left is an audio stream data.